### PR TITLE
System.Text.Json: stackalloc constants + misc PR feedback

### PR DESCRIPTION
**Pull request by CodeBlanch**
Follow-up feedback from #54186

See:

* https://github.com/dotnet/runtime/pull/54186#discussion_r661724474
* https://github.com/dotnet/runtime/pull/54186#pullrequestreview-701223476

/cc @layomia

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs [hunk 1]
```diff
@@ -122,7 +122,7 @@ private void WriteStringEscapeProperty(ReadOnlySpan<char> propertyName, int firs
                     // Cannot create a span directly since it gets assigned to parameter and passed down.
                     unsafe
                     {
-                        char* ptr = stackalloc char[length];
+                        char* ptr = stackalloc char[JsonConstants.StackallocThreshold];
```
Comment (CodeBlanch) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs:125]: These ones using a pointer, is it safe to do like this?

**Comment by GrabYourPitchforks** [comment]
The pattern `(stackalloc byte[...]).Slice(...)` shouldn't be necessary for things which are scratch buffers and where the actual number of bytes written will be received as a variable a few lines below. In these cases, the easiest thing to do would be to `stackalloc` without slicing, perform the operation, then slice the scratch buffer _after_ you know exactly how many bytes were written to it.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs:125]: Yup! But change the line below to read `new Span<char>(ptr, const)` instead of `(..., length)` so that everything matches.

Comment (CodeBlanch) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs:125]: Updated

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Bytes.cs [hunk 2]
```diff
@@ -140,7 +140,7 @@ private void WriteBase64EscapeProperty(ReadOnlySpan<char> propertyName, ReadOnly
             int length = JsonWriterHelper.GetMaxEscapedLength(propertyName.Length, firstEscapeIndexProp);
 
             Span<char> escapedPropertyName = length <= JsonConstants.StackallocThreshold ?
-                stackalloc char[length] :
+                stackalloc char[JsonConstants.StackallocThreshold] :
```
Comment (Tornhoof) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Bytes.cs:143]: Are we happy with using the same threshold for chars, effectivly doubling the stackalloc size? I know the thresholds are all over the place in this repo (100,256,512,1024), but I'm still a bit uneasy with 512 bytes (not even talking about 1024 in MultiPartcontent).

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs [hunk 3]
```diff
@@ -38,7 +38,7 @@ public static byte[] GetEscapedPropertyNameSection(ReadOnlySpan<byte> utf8Value,
             int length = JsonWriterHelper.GetMaxEscapedLength(utf8Value.Length, firstEscapeIndexVal);
 
             Span<byte> escapedValue = length <= JsonConstants.StackallocThreshold ?
-                stackalloc byte[length] :
+                stackalloc byte[JsonConstants.StackallocThreshold] :
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:41]: Was there a request to change from `length` to `StackallocThreshold`? It seems `length` would be more correct unless I'm missing some perf (256 faster than specific number?) or other reason.

Comment (GrabYourPitchforks) [reply src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:41]: Just part of the general "stackallocing a const value is optimized over using a variable value" work.

Comment (CodeBlanch) [reply src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:41]: The original discussion: https://github.com/dotnet/runtime/pull/54186#discussion_r661724474

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Bytes.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Bytes.cs:143]: I'd prefer if the threshold were a fixed number of bytes, scaled accordingly for chars.  That said, as you say we're very inconsistent; we've talked about going through and standardizing everything across the repo, but haven't done so.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Bytes.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Bytes.cs]: Updated to have a scaled threshold for char allocations.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs [hunk 4]
```diff
@@ -54,7 +54,7 @@ public override TimeSpan Read(ref Utf8JsonReader reader, Type typeToConvert, Jso
                 int backslash = source.IndexOf(JsonConstants.BackSlash);
                 Debug.Assert(backslash != -1);
 
-                Span<byte> sourceUnescaped = stackalloc byte[source.Length];
+                Span<byte> sourceUnescaped = stackalloc byte[MaximumEscapedTimeSpanFormatLength];
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs:57]: Can't comment on the exact location, but shouldn't line 33 on this file also be changed to 
```csharp
Span<byte> stackSpan = stackalloc byte[MaximumEscapedTimeSpanFormatLength];
```

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs:57]: Shouldn't this be
```suggestion
                Span<byte> sourceUnescaped = stackalloc byte[MaximumTimeSpanFormatLength];
```
given that it's writing the unescaped form?

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonDateTimeTestData.cs [hunk 5]
```diff
@@ -232,6 +232,11 @@ public static IEnumerable<object[]> InvalidISO8601Tests()
             // High byte expansion - parsing fails early at 254 characters.
             yield return new object[] { "\"" + new string('\u20AC', 250) + "\"" };
             yield return new object[] { "\"" + new string('\u20AC', 260) + "\"" };
+
+            // Whitespace
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonDateTimeTestData.cs:236]: Why was this added, out of curiosity?

Comment (CodeBlanch) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonDateTimeTestData.cs:236]: Just to be consistent. There was a line break between the other "sections" in the method except for that one. I can revert if you want.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs:57]: * Line 38: This originally came from https://github.com/dotnet/runtime/pull/54186#discussion_r662533772. Happy to change if you want. Not sure what will have best perf, checking the flag or allocating MaximumEscapedTimeSpanFormatLength (x3) when we know MaximumTimeSpanFormatLength is all we need?

* Line 57: I made the change but the tests blow up. There's this assert in Unescape:

https://github.com/dotnet/runtime/blob/188c2702cd6f365bfad3a985f82e1f9cd520aeb4/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L327

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs:57]: - Line 38: we can leave it be for now and make a call later.
- Line 57: I might have misinterpreted how that particular buffer is being used. Sorry about that.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks. I rebased your changes to fix the merge conflict.
